R/process_regions.R

```diff
@@ -0,0 +1,123 @@
+#' @param projection Specify which genome build to use. [grch37]
+#'
+#' @return Numeric value.
+#'
+#' @noRd
+#'
+#' @examples
+#' regions <- c("chr18:63074150-63369324")
+#' process_regions(regions)
+#' 
+#' these_regions <- process_regions(
+#'  only_regions = c("MYC", "BCL2", "BCL6")
+#' )
+#' regions_bed <- these_regions$regions_bed
+#' regions_vec <- these_regions$regions
+#' @export
+process_regions <- function(regions_list = NULL,
+                            regions_bed = NULL,
+                            region_padding = 0,
+                            skip_regions = NULL,
+                            only_regions = NULL,
+                            projection = "grch37") {
+  # Use default ashm region table if no regions are provided
+  if (is.null(regions_list)) {
+    if (is.null(regions_bed)) {
+      message("Using default GAMBLR aSHM regions. ")
+      if (projection == "grch37") {
+        regions_bed <- GAMBLR.data::grch37_ashm_regions %>%
+          dplyr::mutate(chr_name = str_remove(chr_name, "chr")) %>% 
+          dplyr::mutate(name = str_c(gene, region, sep = "_"))
+      } else {
+        regions_bed <- GAMBLR.data::hg38_ashm_regions %>%
+          dplyr::mutate(name = str_c(gene, region, sep = "_"))
+      }
+      # Fix column names
+      regions_bed <- regions_bed %>%
+        rename_with(
+          ~ str_remove(.x, "^hg.*_")
+        ) %>%
+        dplyr::rename(chrom = chr_name)
+      if (!is.null(skip_regions)) {
+        # drop user-specified regions
+        regions_bed <- regions_bed %>%
+          dplyr::filter(!gene %in% skip_regions)
+      }
+      if (!is.null(only_regions)) {
+        # keep only user-specified regions
+        regions_bed <- regions_bed %>%
+          dplyr::filter(gene %in% only_regions)
+      }
+    }
+    
+    required_cols <- c("chrom", "start", "end", "name")
+    if (min(required_cols %in% colnames(regions_bed)) == 0) {
+      stop("Provided regions_bed lacks required column names. Ensure columns chrom, start, end, and name are present. ")
+    }
+    
+    # gene column is required for later joins
+    if (!"gene" %in% colnames(regions_bed)) {
+      regions_bed <- mutate(regions_bed, gene = name)
+    }
+  } else {
+    # Convert character vector of regions to df
+    regions_bed <- bind_rows(lapply(regions_list, function(x) {
+      chunks <- region_to_chunks(x)
+      df <- data.frame(
+        chrom = chunks$chromosome,
+        start = as.numeric(chunks$start),
+        end = as.numeric(chunks$end)
+      )
+    }))
+    if (!is.null(names(regions_list))) {
+      regions_bed$name <- names(regions_list)
+      regions_bed$gene <- names(regions_list)
+    } else {
+      message("WARNING: Regions provided as an unnamed character vector. It is strongly recommended to provide a named vector with a unique name per region, or a bed file-formatted data frame. ")
+      regions_bed$name <- regions_bed$chrom
+      regions_bed$gene <- regions_bed$chrom
```

Comment:
I'm not sure whether the function returns the desired output when `regions_list` is unnamed.
If there are no names, the function uses the chromosome name as the region name. If regions are from the same chromosome, they are merged (_"...based on min(start) and max(end)..."_). Besides, it's confusing when one warning message says `Regions provided as an unnamed...` and the other `... have the same name`. 
```
> regions <- c("chr1:10000-15000", "chr1:100000000-100005000")
> process_regions(regions_list = regions)
WARNING: Regions provided as an unnamed character vector. It is strongly recommended to provide a named vector with a unique name per region, or a bed file-formatted data frame. 
Warning: Multiple regions in the provided data frame have the same name. Merging these entries based on min(start) and max(end) per name value. 
$regions_list
                  chr1 
"chr1:10000-100005000" 

$regions_bed
# A tibble: 1 × 5
  chrom start       end name  gene 
  <chr> <dbl>     <dbl> <chr> <chr>
1 chr1  10000 100005000 chr1  chr1 
```

when the chromosomes are different, the regions are not merged:
```
> regions <- c("chr1:10000-15000", "chr2:100000000-100005000")
> process_regions(regions_list = regions)
WARNING: Regions provided as an unnamed character vector. It is strongly recommended to provide a named vector with a unique name per region, or a bed file-formatted data frame. 
$regions_list
                  chr1                   chr2 
    "chr1:10000-15000" "chr2:1e+08-100005000" 

$regions_bed
  chrom start       end name gene
1  chr1 1e+04     15000 chr1 chr1
2  chr2 1e+08 100005000 chr2 chr2
```

Reply:
The function strongly recommends using names.

Reply:
I can have a look at this, thanks for reporting!